ssl/t1_lib.c

```diff
@@ -1021,167 +1096,604 @@ int tls1_set_groups(uint16_t **pext, size_t *pextlen,
-    gcb.ctx = ctx;
-    if (!CONF_parse_list(str, ':', 1, gid_cb, &gcb))
+
+    /*
+     * Start the (potentially recursive) tuple processing by calling CONF_parse_list
+     * with the TUPLE_DELIMITER_CHARACTER (which will call tuple_cb after cleaning spaces)
+     */
+    parse_ret = CONF_parse_list(str, TUPLE_DELIMITER_CHARACTER, 1, tuple_cb, &gcb);
+
+    if (parse_ret == 0)
         goto end;
+    if (parse_ret == -1) {
+        ERR_raise_data(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT,
+                       "Syntax error in '%s'", str);
+        goto end;
+    }
+
+    /*
+     * We check whether a tuple was completly emptied by using "-" prefix excessively,
+     * in which case we remove the tuple
+     */
+    for (i = 0; i < gcb.tplcnt; i++) {
+        if (gcb.tuplcnt_arr[i] == 0) {
+            for (j = i; j < (gcb.tplcnt - 1); j++) /* Move tuples to the left */
+                gcb.tuplcnt_arr[j] = gcb.tuplcnt_arr[j + 1];
+
+            gcb.tplcnt--; /* We just deleted a tuple, update book keeping */
+            i--; /* Acount for the fact that the list is shorter now */
+        }
+    }
+
+    /* Some more checks (at least one remaining group, not more that nominally 4 key shares */
     if (gcb.gidcnt == 0) {
         ERR_raise_data(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT,
                        "No valid groups in '%s'", str);
         goto end;
     }
-    if (pext == NULL) {
+    if (gcb.ksidcnt > OPENSSL_CLIENT_MAX_KEY_SHARES) {
+        ERR_raise_data(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT,
+                       "To many keyshares requested in '%s' (max = %d)",
+                       str, OPENSSL_CLIENT_MAX_KEY_SHARES);
+        goto end;
+    }
+
+    /*
+     * For backward compatibility we let the rest of the code know that a key share
+     * for the first valid group should be added if no "*" prefix was used anywhere
+     */
+    if (gcb.ksidcnt == 0) {
+        /*
+         * No key share group prefix character was used, hence we indicate that a single
+         * key share should be sent and flag that it should come from the supported_groups list
+         */
+        gcb.ksidcnt = 1;
+        gcb.ksid_arr[0] = 0;
+    }
+
+    /*
+     * A call to tls1_set_groups_list with any of the args (other than ctx) set
+     * to NULL only does a syntax check, hence we're done here and report success
+     */
+    if (grpext == NULL || ksext == NULL || tplext == NULL ||
+        grpextlen == NULL || ksextlen == NULL || tplextlen == NULL) {
         ret = 1;
         goto end;
     }
 
     /*
-     * gid_cb ensurse there are no duplicates so we can just go ahead and set
-     * the result
+     * tuple_cb and gid_cb combo ensures there are no duplicates or unknown groups so we
+     * can just go ahead and set the results
      */
-    tmparr = OPENSSL_memdup(gcb.gid_arr, gcb.gidcnt * sizeof(*tmparr));
-    if (tmparr == NULL)
+    repl_grpext = OPENSSL_memdup(gcb.gid_arr, gcb.gidcnt * sizeof(*repl_grpext));
```

Comment:
I'm wondering why we memdup the `gcb.gid_arr` buffer rather than just taking it directly. We end up just freeing it below anyway, e.g.

```suggestion
    repl_grpext = gcb.gid_arr;
    gcb.gid_arr = NULL;
```

Reply:
I hadn't really noticed that this pattern was existing. I just spotted something I thought was slightly odd. I think my proposed approach makes things cleaner (but its also possible I missed some subtlety). But I wouldn't insist on it, and will leave it to your discretion as to whether to accept my proposal or ignore it.

Reply:
Avoiding any dynamic memory [de-]allocation which is not absolutely required is a good thing.
--> Will be changed in the next update to directly use the available memory.

Minor caveat: The memory allocated for e.g. `gcb.gid_arr` is driven by (multiples of) `GROUPLIST_INCREMENT` (= 32), while the `OPENSSL_memdup` is only duplicating the minimum required bytes. 
Given the fairly small size of `GROUPLIST_INCREMENT`, I don't think that is any issue, but wanted to explicitly mention it.

Reply:
Yes, of course. All now treated the same way:
```
    OPENSSL_free(*grpext);
    *grpext = gcb.gid_arr;
    *grpextlen = gcb.gidcnt;

    OPENSSL_free(*ksext);
    *ksext = gcb.ksid_arr;
    *ksextlen = gcb.ksidcnt;

    OPENSSL_free(*tplext);
    *tplext = gcb.tuplcnt_arr;
    *tplextlen = gcb.tplcnt;
```
followed by a `return 1;` such that the `gcb.xyz_arr` memory is not deallocated at `end:`.